R/scVI.R

```diff
@@ -28,63 +35,149 @@ NULL
+    adata = adata,
+    n_latent = as.integer(x = ndims),
+    n_layers = as.integer(x = nlayers),
+    gene_likelihood = gene_likelihood
+  )
   model$train(max_epochs = max_epochs)
-  latent = model$get_latent_representation()
+
+  # extract the latent representation of the merged data
+  latent <- model$get_latent_representation()
   latent <- as.matrix(latent)
+  # pull the cell identifiers back out of the `AnnData` instance
+  # in case anything was sorted under the hood
   rownames(latent) <- reticulate::py_to_r(adata$obs$index$values)
+  # prepend the latent space dimensions with `new.reduction` to
+  # give the features more readable names
   colnames(latent) <- paste0(new.reduction, "_", 1:ncol(latent))
-  suppressWarnings(latent.dr <- CreateDimReducObject(embeddings = latent, key = new.reduction))
+
+  # build a `DimReduc` instance
+  suppressWarnings(
+    latent.dr <- CreateDimReducObject(
+      embeddings = latent, 
+      key = new.reduction
+    )
+  )
+  # to make it easier to add the reduction into a `Seurat` instance
+  # we'll wrap it up in a named list
   output.list <- list(latent.dr)
   names(output.list) <- new.reduction
+
   return(output.list)
 }
 
-attr(x = scVIIntegration, which = 'Seurat.method') <- 'integration'
+attr(x = scVIIntegration, which = "Seurat.method") <- "integration"
+
+
+#' Builds a data.frame with batch identifiers to use when integrating
+#' \code{object}. For \code{SCTAssay}s, batches are split using their
+#' model identifiers. For \code{StdAssays}, batches are split by layer.
+#'
+#' This function is essentially the same as \code{Seurat:::CreateIntegrationGroups}
+#' except that it does not take in a `scale.layer` param.
+#'
+#' @param object A \code{SCTAssay} or \code{StdAssays} instance.
+#' @param layers Layers in \code{object} to integrate.
+#'
+#' @return A dataframe indexed on the cell identifiers from \code{object} - 
+#' the dataframe contains a single column, "batch", indicating the ...
+.FindBatches <- function(object, layers) {
+  # if an `SCTAssay` is passed in it's expected that the transformation
+  # was run on each batch individually and then merged so we can use
+  # the model identifier to split our batches
+  if (inherits(object, what = "SCTAssay")) {
+    # build an empty data.frame indexed
+    # on the cell identifiers from `object`
+    batch.df <- SeuratObject::EmptyDF(n = ncol(object))
+    row.names(batch.df) <- Cells(object)
+    # for each
+    for (sct.model in levels(object)) {
+      cell.identifiers <- Cells(object, layer = sct.model)
+      batch.df[cell.identifiers, "batch"] <- sct.model
+    }
+    # otherwise batches can be split using `object`'s layers
+  } else {
+    # build a LogMap indicating which layer each cell is from
+    layer.masks <- slot(object, name = "cells")[, layers]
+    # get a named vector mapping each cell to its respective layer
+    layer.map <- labels(
+      layer.masks,
+      values = Cells(object, layer = layers)
```

Comment:
@mhkowalski would you mind giving this function a sanity check for me? 

Since `CreateIntegrationGroups` isn't exported by `Seurat` it seemed simpler to just re-implement the logic here. I've dropped the `scale.layer` param since. 

**As far as I can tell** this was just being used get the cell IDs for all of the layers since there's only ever one scale layer present. Unless the order is important here, or there's a possibility that the data is subset during normalization/scaling than I think this should be safe (especially since `scVIIntegration` itself forces the use of the "counts" layers). 

I might be missing something though...